Fixes bug introduced by #3596 when handling NaNs.

pepbos opened the pull request:
My Pr #3596 introduced a bug, or triggered a bug elsewhere, when handling NaN's in SmoothSegmentedFunction.

Not sure when NaNs will be injected, and if that even makes sense, but before #3596 that was OK, but currently main will raise an exception. This PR makes it fine again, as was before #3596.

I noticed when trying to visualize my result from a CMC, which raised an exception now.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/3600)
<!-- Reviewable:end -->

nickbianco commented on OpenSim/Common/SmoothSegmentedFunction.cpp:
```diff
@@ -584,50 +584,52 @@ ________________________________________________________________________
-            ctrlPtsY[idx],
-            order);
+    // In case of NaN return zero.
+    return 0.;
```
Do you think this might be hiding another issue? Is it possible that your CMC simulation/visualization is feeding this function a NaN value for `x` which then gets masked by returning zero here. Or do you think the changes introduced new NaNs somewhere?

nickbianco replied:
I see, this was the previous behavior of the function. In that case, this is probably fine.

pepbos commented:
Looks like the intention was for the smoothSegmentedFunction to return NaN for a NaN as input (which makes sense).
However it had three different cases:
- `calcValue(NaN)` returned `NaN` 
- `calcDerivative(NaN, 1)` returned `0` 
- `calcValue(NaN, 2)` returned `dydx1`

Changed this PR to fix always returning `NaN` when the input is `NaN`, so:
- `calcValue(NaN)` return `NaN` 
- `calcDerivative(NaN, 1)` return `NaN` 
- `calcDerivative(NaN, 2)` return `NaN` 

Note that without this PR an exception is raised if the input is NaN, which blows up some cases that were fine before.
I added the NaN case to the unittests.

nickbianco reviewed (approved):
LGTM. Thanks for the clarification @pepbos.

nickbianco commented:
Also, you can ignore the Mac failures, that is an issue with the CI. Working on a fix for that.
